shim/src/data/send.c

```diff
@@ -43,10 +43,36 @@ ssize_t __send(int sockfd, const void *buf, size_t len, int flags)
     len = MIN(len, sga.sga_segs[0].sgaseg_len);
     memcpy(sga.sga_segs[0].sgaseg_buf, buf, len);
     assert(__demi_push(&qt, sockfd, &sga) == 0);
-    assert(__demi_wait(&qr, qt, NULL) == 0);
-    assert(qr.qr_opcode == DEMI_OPC_PUSH);
-    __demi_sgafree(&sga);
 
+    // We need to add the send to a special epoll table that gets
+    // polled at epoll_wait. This is because demi_wait
+    // only returns from a pop when the demikernel receives an ACK back.
+    // Doing the wait here messes up the behaviour of applications
+    // that expect the send to return immediately after the OS adds
+    // the buffer to the transmit queue.
+    for (int i = 0; i < MAX_EVENTS; i++)
+    {
+        struct demi_event *ev = epoll_get_event(SEND_EPFD, i);
+        if (ev->used == 0)
+        {
+            ev->used = 1;
+            ev->qt = qt;
+            ev->sockqd = sockfd;
+            ev->next_ev = INVALID_EV;
+            ev->prev_ev = INVALID_EV;
+            ev->sga = sga;
```

Comment:
Do we have this code snippet all around? I feel that we should have an initializer function for that if we are always initializing with some default values.